data/src/main/java/com/microsoft/azure/kusto/data/auth/TokenProviderFactory.java

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.kusto.data.auth;
+
+import com.microsoft.aad.msal4j.ClientCredentialFactory;
+import com.microsoft.aad.msal4j.IClientCertificate;
+import com.microsoft.aad.msal4j.IClientSecret;
+import org.apache.commons.lang3.StringUtils;
+import org.jetbrains.annotations.NotNull;
+
+import java.net.URISyntaxException;
+import java.util.concurrent.Callable;
+
+public class TokenProviderFactory {
+    private TokenProviderFactory() {
+        // Hide constructor as this is a Factory
+    }
+
+    public static TokenProviderBase createTokenProvider(@NotNull ConnectionStringBuilder csb) throws URISyntaxException {
+        String clusterUrl = csb.getClusterUrl();
+        String authorityId = csb.getAuthorityId();
+        if (StringUtils.isNotBlank(csb.getApplicationClientId()) && StringUtils.isNotBlank(csb.getApplicationKey())) {
+            String clientId = csb.getApplicationClientId();

```

Comment:
nit:
Since you do not need else clauses due to return on success,
```
load params 
if (checks)
{
 return
}

load next parameter set
...
```

Reply:
Different strokes for different folks -
I always cringe when I see that design because I think the "else"s add significantly to readability (similar to  https://stackoverflow.com/a/3261899).

My philosophy: The code is more for the programmer to read than the compiler.

Reply:
Here's another interesting discussion about this, which I very much agree with: https://blog.mozilla.org/nnethercote/2009/08/31/no-else-after-return-considered-harmful/

Reply:
Your call, 
just note you make double calls to property getters and your code is much more verbose then it should be.
Readability-wise you loose points here